gobblin-runtime/src/main/java/org/apache/gobblin/runtime/Task.java

```diff
@@ -364,8 +365,12 @@ public void run() {
     } catch (Throwable t) {
       failTask(t);
     } finally {
-      this.taskStateTracker.onTaskRunCompletion(this);
-      completeShutdown();
+      if (this.taskFuture == null || !this.taskFuture.isCancelled()) {
```

Comment:
This block needs to synchronize access to the taskFuture with the cancel() call below. It should also modify the state of the taskFuture to avoid double cancellation when the cancel() block below gets called after this block executes.

Reply:
I will add synchronized access to the future, but I think changing future's state is not required. The state should be changed to 'completed' when it actually completes.